### ISSUE TITLE
Would disabling FLAC CRC checks help fuzzing

### ISSUE THREAD

**Issue by ktmf01** [google/oss-fuzz#7132]
Not really an issue but more of a question really: would it help if the FLAC configure script enabled a way to disable the CRC checking in the headers? I guess it would, but in the past I've ran afl on libFLAC code and didn't see much of a difference in the code coverage when disabling CRC checks, but still I think it should.

I've been sending some PRs to the FLAC project with fixes for problems oss-fuzz found, but especially if at some point CIFuzz is added to FLAC, this might make all the difference.

Please let me know what you think.

**Comment by jonathanmetzman** [comment]
I would assume this would help

>I guess it would, but in the past I've ran afl on libFLAC code and didn't see much of a difference in the code coverage when disabling CRC checks, but still I think it should.

Surprising this didn't help. Fuzzing is all about things in my head not matching reality though.

>but especially if at some point CIFuzz is added to FLAC, this might make all the difference.

Nice. What's stopping this?

**Comment by ktmf01** [comment]
> configure script enabled a way to disable the CRC checking

I see now there is a preprocessor macro that is being used by oss-fuzz: FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION I just used that define twice in libFLAC code and it results in much higher cov and ft number, but much lower speed. Perhaps I did something wrong when I checked with AFL.

Here are some abridged fuzzer outputs:
>Current flac repository
1e8:  cov 89 ft 343
3e8:  cov 89 ft 347
5e8:  cov 89 ft 351
exec/s: 2e5
>
>With crc disabled through FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION
1e8:  cov 201 ft 1102
3e8:  cov 201 ft 1108
5e8:  cov 201 ft 1134
exec/s: 6e4

So, fuzzing is 3 times as slow (presumably because more code is run each fuzz try), but coverage is more than doubled. I'll send in a PR to flac, but probably won't be merged soon.

> Nice. What's stopping this?

libFLAC is apparently short on maintainers with time to spare. Last change to the repo was half a year ago, and I have 16 open PRs for it, including fixing a few oss-fuzz finds, of which one is already more than 2 years old. Integration of oss-fuzz a few years ago was also mostly driven "externally".

**Comment by ktmf01** [comment]
I'm commenting once again for any people that may find this through Google or something.

I did a thorough comparison which took two weeks to complete. I ran a fuzzer with a known problem with and without these CRC checks. In this case, only an 8-bit CRC is blocking further processing, a 16-bit CRC is only checked after running through all code. The 16-bit CRC might still confuse the fuzzer though.

Running the fuzzer 8 times with 8-bit CRC checking disabled resulted in the following number of execs to find the problem
Min: 2.042.264
Max: 13.823.448

Further disabling the 16-bit CRC and padding bit checks gave the following results for another 8 runs
Min: 542.806
Max: 11.085.011

Not disabling the 8-bit CRC confused the fuzzer so badly, that I had to set an upper limit to the number of execs. I set the limit to 2 billion execs, which amounts to about 4 days of single-thread performance on my fairly recent server, but half of the runs this limit was reached. The minimum number of execs seen was 4.208.446, but a maximum could not be established

One could say disabling the 8-bit CRC therefore improved fuzzing speed mostly by not having the fuzzer get stuck. Time to find the problem is brought back from several days to several tens of minutes, but the real gain depends on what limit is set to the time the fuzzer is run.
